system/codesigner/gpgcodesigner.cpp

```diff
@@ -0,0 +1,278 @@
+    int ret = runExternalCommand(output, errmsg, "gpg --list-secret-keys --with-colon", nullptr);
+    if (ret != 0)
+        throw MakeStringException(-1, "Error running gpg: %s", errmsg.str());
+
+    const char* START = "\nuid:";
+    if (isGpgV1)
+        START = "\nsec:";
+    int startlen = strlen(START);
+    const int SKIP = 8;
+    const char* line = output.str();
+    StringArray uids;
+    while (line && *line)
+    {
+        line = strstr(line, START);
+        if (!line)
+            break;
+        line += startlen;
+        line = skipn(line, ':', SKIP);
+        if (!*line)
+            break;
+        const char* uid_s = line;
+        while (*line != '\0' && *line != ':')
+            line++;
+        if (line > uid_s)
+        {
+            StringBuffer uid(line - uid_s, uid_s);
+            uid.trim();
+            if (uid.length() > 0)
+                uids.append(uid.str());
+        }
+    }
+    uids.sortAscii(false);
+    const char* current = "";
+    for (unsigned i = 0; i < uids.length(); i++)
+    {
+        if (strcmp(uids.item(i), current) != 0)
+        {
+            current = uids.item(i);
+            userIds.append(current);
+        }
+    }
+
+    return userIds;
+}
+
+bool GpgCodeSigner::getKeyGrip(const char *userId, StringBuffer &keygrip)
+{
+    initGpg();
+    keygrip.clear();
+
+    StringBuffer cmd;
+    if (isGpgV1)
+        cmd.appendf("gpg --list-secret-keys \"=%s\"", userId); // = means exact match
+    else
+        cmd.appendf("gpg --list-secret-keys --with-keygrip \"=%s\"", userId); // = means exact match
+
+    StringBuffer output, errmsg;
+    int ret = runExternalCommand(output, errmsg, cmd.str(), nullptr);
+    if (ret != 0 || strstr(output.str(), userId) == nullptr)
+        return false;
+    auto kgptr = strstr(output.str(), "Keygrip = ");
+    if (kgptr)
+        keygrip.append(40, kgptr+10);
+    else
+        return false;
+    return true;
+}
+
+void GpgCodeSigner::clearPassphrase(const char* key)
+{
+    initGpg();
+    StringBuffer output, errmsg;
+    VStringBuffer cmd("gpg-connect-agent \"clear_passphrase --mode=normal %s\" /bye", key);
+    runExternalCommand(output, errmsg, cmd.str(), nullptr);
+}
+
+StringBuffer & GpgCodeSigner::sanitizeUserId(const char * userId, StringBuffer & sanitizedUserId)
+{
+    sanitizedUserId.clear();
+    sanitizedUserId.ensureCapacity(strlen(userId));
+    const char *p = userId;
+    while (*p && *p==' ') ++p;
+    while (*p)
+    {
+        if (isalnum(*p) || *p==' ' || *p=='.' || *p=='@' || *p =='<' || *p=='>')
+            sanitizedUserId.append(*p);
+        ++p;
+    };
+    return sanitizedUserId;
+}
+
+StringBuffer & GpgCodeSigner::sanitizePassphrase(const char * passphrase, StringBuffer & sanitizedPassphrase)
+{
+    sanitizedPassphrase.clear();
+    sanitizedPassphrase.ensureCapacity(strlen(passphrase));
+    const char *p = passphrase;
+    // Note: strictly speaking any characters should be valid for passphrases. However, fully
```

Comment:
Would it be better to escape the chars that could possibly cause injection? Another option might be to include the whole argument inside quotes, and escape the quote chars maybe?

Reply:
The user id is already quoted.  We can't quote the passphrases.  We could either filter our passphrases containing quotes, control characters or too long?

Reply:
I took a closer look. Passphrase is directly streamed to std input and it doesn't appear any where on the command line, so we don't need to filter it. The sanitize method silently drops chars, that could be a cause of confusion for users. So basically we can just leave the passphrase as it's passed in I think.

Reply:
Removed.